app/models/attachment_reference.rb

```diff
@@ -0,0 +1,61 @@
+class AttachmentReference < ActiveRecord::Base
+  after_initialize :set_expiry_at, if: proc { new_record? && !attachable }
+  before_validation :update_expiry_at
```

Comment:
Why is the record always valid?

Creating a new record with attributes will call `attachable=`; try it within a debugger. All attributes in the initializer will be transformed to a attribute= method and those will be called.